src/AuditTrailModule/javasource/audittrail/log/CreateLogObject.java

```diff
@@ -296,115 +292,141 @@ private static IMendixObject createSingleLogLine(IMendixObject logObject, IMendi
 						(Integer) logObject.getValue(context, Log.MemberNames.NumberOfChangedMembers.toString()) + 1);
 			}
 
-			return logLine;
+			return Collections.singletonList(logLine);
 		}
 
 		_logNode.trace("Skipping member: " + member.getName() + " because it has not changed.");
-		return null;
+		return Collections.emptyList();
 	}
 
-	private static IMendixObject createReferenceLogLine(IMendixObject logObject, IMendixObjectMember<?> member,
+	private static List<IMendixObject> createReferenceLogLine(IMendixObject logObject, IMendixObjectMember<?> member,
 			boolean isNew, IContext sudocontext, IContext currentcontext) throws CoreException {
 		// get current and previous id
 		IMendixIdentifier cID = (IMendixIdentifier) member.getValue(currentcontext);
 		IMendixIdentifier pID = (IMendixIdentifier) member.getOriginalValue(currentcontext);
 
-		// Get the values of reference objects
-		String pValue = getValueFromReference(pID, currentcontext);
-		String newValue = getValueFromReference(cID, currentcontext);
-
-		if (IncludeOnlyChangedAttributes == false || !pValue.equals(newValue) || isNew) {
+		if (IncludeOnlyChangedAttributes == false || !Objects.equals(cID, pID) || isNew) {
+			List<IMendixObject> logLineList = new ArrayList<IMendixObject>();
 			IMendixObject logLine = Core.instantiate(sudocontext, LogLine.getType());
 
 			logLine.setValue(sudocontext, LogLine.MemberNames.Member.toString(), member.getName());
 			logLine.setValue(sudocontext, LogLine.MemberNames.MemberType.toString(), MemberType.Reference.toString());
 			logLine.setValue(sudocontext, LogLine.MemberNames.LogLine_Log.toString(), logObject.getId());
-			logLine.setValue(sudocontext, LogLine.MemberNames.NewValue.toString(), newValue);
-
-			if (isNew)
-				logLine.setValue(sudocontext, LogLine.MemberNames.OldValue.toString(), "");
-			else
-				logLine.setValue(sudocontext, LogLine.MemberNames.OldValue.toString(), pValue);
+			logLine.setValue(sudocontext, LogLine.MemberNames.NewValue.toString(), "");
+			logLine.setValue(sudocontext, LogLine.MemberNames.OldValue.toString(), "");
+
+			logLineList.add(logLine);
+			
+			if (cID == pID && cID != null) {
+				logLineList.addAll(createLogLinesForReferencedObject(pID, logLine.getId(), currentcontext, sudocontext, TypeOfReferenceLog.No_Change));
+			} else {
+				if (cID != null) {
+					logLineList.addAll(createLogLinesForReferencedObject(cID, logLine.getId(), currentcontext, sudocontext, TypeOfReferenceLog.Deleted));

```

Comment:
I am a bit confused about these checks. When the current id is _**null**_ doesn't that mean that the object is deleted?

Reply:
It's confusing because it's wrong. 😅